### PR TITLE
issue #968 Drag & Drop limit of 100 files/folders

### PR DESCRIPTION
**Pull request by lmcbout**
Allow to read all files within a folders even when there is
more than 100 files in it.

Signed-off-by: Jacques Bouthillier <jacques.bouthillier@ericsson.com>

### CODE REVIEW

**Comment by lmcbout** [comment]
Ready to be reviewed

File: packages/filesystem/src/browser/file-tree/file-tree-model.ts [hunk 1]
```diff
@@ -148,7 +148,25 @@ export class FileTreeModel extends TreeModel implements LocationService {
...
+        this.readEntries(reader, newBase);
+    }
+
+    protected readEntries(reader: WebKitDirectoryReader, dest: URI): void {
+        let allFiles = [] as WebKitFileEntry[];
+        const getEntries = () => {
+            reader.readEntries((results: any) => {
+                if (results.length) {
+                    allFiles = allFiles.concat(results);
+                    getEntries(); // loop to read all entries
```
Comment (akosyakov) [inline packages/filesystem/src/browser/file-tree/file-tree-model.ts]: Could you elaborate why it is needed? The code looks very tricky to grasp.

**Comment by lmcbout** [comment]
When there are more than 100 files or folders within the same folder,  we currently only read the first 100 items and forget everything else. With the new code, we read all available files or folders, fill the allowed buffer of max 100 entries and read the folder again until we processed all entries available within this folder, then we proceed to send the content the browser to display it. So if for example there is 150 files in the folders to process, we read the first iteration, get the first 100 entries, keep it in a local buffer, read again to get the last 50 entries, then we process all entries to the browser.. So all data is as expected.

File: packages/filesystem/src/browser/file-tree/file-tree-model.ts [hunk 2]
```diff
@@ -148,7 +148,25 @@ export class FileTreeModel extends TreeModel implements LocationService {
         if (!await this.fileSystem.exists(uri)) {
             await this.fileSystem.createFolder(uri);
         }
-        reader.readEntries(items => this.uploadEntries(newBase, items as any));
+        this.readEntries(reader, newBase);
+    }
+
+    protected readEntries(reader: WebKitDirectoryReader, dest: URI): void {
+        let allFiles = [] as WebKitFileEntry[];
```
Comment (akosyakov) [inline packages/filesystem/src/browser/file-tree/file-tree-model.ts]: Do you need it? Will it be possible to upload in the reader callback?

Comment (lmcbout) [reply packages/filesystem/src/browser/file-tree/file-tree-model.ts]: Added a comment to the readEntries method,
Put the upload within the callback.

**Comment by lmcbout** [comment]
Ready to be reviewed , comment added

**Comment by akosyakov** [comment]
@lmcbout looks good, I've only split dir reading from uploading if tests are green please merge

**Comment by lmcbout** [comment]
Code re based and merged

**Pull request merged**